### PR TITLE
Tighten CLI logging and jq option validation

### LINKED ISSUE #58: Make SQL/DML classification robust to comments and statement prefixes
## Problem

DML mode selection currently depends on a regex that only matches statements beginning with `INSERT`, `UPDATE`, or `DELETE` after whitespace:

```go
var dmlRe = regexp.MustCompile(`(?is)^\s*(INSERT|UPDATE|DELETE)\s+.+$`)
```

Valid DML with leading SQL comments or statement prefixes can be routed through the read-only query path, producing confusing runtime errors instead of using the read-write transaction path.

## Examples to cover

```sql
-- comment
UPDATE Singers SET FirstName = 'x' WHERE SingerId = 1
```

```sql
/* comment */
DELETE FROM Singers WHERE SingerId = 1
```

## Verified locally

Confirmed on `main` at `4d2abf6`: `dmlRe.MatchString(query)` is the sole DML classifier in `_main`.

## Suggested fix

Replace the regex-only check with a small classifier that strips leading whitespace and SQL comments, then checks the first keyword. Keep `--enable-partitioned-dml` as the explicit override.

## Acceptance criteria

- Add table-driven tests for DML/non-DML classification.
- DML with leading comments routes to read-write execution.
- SELECT statements remain read-only.

### LINKED ISSUE #59: Redact or further gate gRPC payload logging
## Problem

`--log-grpc` installs payload logging interceptors for all gRPC methods with predicates that always return `true`:

```go
grpczap.PayloadUnaryClientInterceptor(zapLogger, func(...) bool { return true })
grpczap.PayloadStreamClientInterceptor(zapLogger, func(...) bool { return true })
```

This can expose SQL text, query parameters, row values, query plans, and stats in logs. Useful for local debugging, but unsafe in shared terminals, CI logs, support bundles, or production traces.

## Verified locally

Confirmed on `main` at `4d2abf6` in `logGrpcClientOptions()`.

## Suggested fix

Consider one of:

- Change `--log-grpc` to metadata-only logging by default.
- Add `--log-grpc=metadata|payload`.
- Add a separate `--unsafe-log-grpc-payloads` flag for full payloads.
- Redact `Statement.SQL`, `Statement.Params`, and result row values by default.

## Documentation

Update README to state that full gRPC payload logging is sensitive and should not be used in shared CI/support logs unless explicitly intended.

## Acceptance criteria

- Safe default logging mode does not emit parameter values or rows.
- Full payload logging remains available for local debugging behind an explicit unsafe option.

### LINKED ISSUE #60: Honor timestamp bound in --try-partition-query for historical schema checks
## Context

`--try-partition-query` checks whether a query can be executed as a partition query; it does not return row data. Internally it still uses `BatchReadOnlyTransaction.PartitionQuery`, which runs query planning against a read-only snapshot.

For the common case — **“can this query be partitioned against the current schema?”** — always using `StrongRead()` is sufficient.

However, partitionability depends on the **query plan at a specific read timestamp**, which in turn depends on schema state at that time (available indexes, interleaving, `@{FORCE_INDEX=...}` hints, etc.). Spanner retains schema/version history within `version_retention_period`, so checking partitionability at a past timestamp is a valid use case when you want to know whether a query would have been partitionable under a historical schema — for example before/after a DDL change, or to match the timestamp bound you will use when actually running partitioned reads.

## Problem

`_main` computes a timestamp bound `tb` from `--read-timestamp` or `--strong`, but the `--try-partition-query` path always creates a batch read-only transaction with `spanner.StrongRead()`:

```go
bt, err := client.BatchReadOnlyTransaction(ctx, spanner.StrongRead())
```

If the user passes `--read-timestamp`, the flag is silently ignored in this path. That is inconsistent with the rest of the CLI and prevents historical schema checks within the retention window.

This is primarily a **consistency / optional advanced use case** issue, not a breakage of the default “check now” workflow.

## Verified locally

Confirmed on `main` at `4d2abf6`: partition-query check ignores `tb` (around line 360).

## Suggested fix

Use the computed timestamp bound:

```go
bt, err := client.BatchReadOnlyTransaction(ctx, tb)
```

No behavior change when neither `--read-timestamp` nor other staleness flags are set (`tb` remains strong read).

## Documentation

Briefly note in README that:

- default `--try-partition-query` validates against the current schema (strong read)
- `--read-timestamp` can be used to validate against a historical schema within `version_retention_period`

## Acceptance criteria

- `--try-partition-query` with `--read-timestamp` uses the specified timestamp bound.
- `--try-partition-query` without `--read-timestamp` continues to use strong read.
- Add a test (integration or seam) covering the timestamp-bound path if feasible.

### LINKED ISSUE #61: Fix Go 1.25 drift in README and container publish workflow
## Problem

`go.mod` requires Go 1.25.0, but README still says local build requires Go 1.24, and the `ko` workflow pins Go 1.24.x:

```yaml
# .github/workflows/ko.yml
- uses: actions/setup-go@v4
  with:
    go-version: '1.24.x'
```

This can break container publishing and confuse local contributors. Other workflows (`go.yml`, `golangci-lint.yml`) already use `go-version-file: go.mod`.

## Verified locally

- `go.mod`: `go 1.25.0`
- `README.md` line 58: "Local build requires Go 1.24."
- `.github/workflows/ko.yml`: `go-version: '1.24.x'`

## Suggested fix

- Update README to say Go 1.25.
- Change `ko.yml` to use:

```yaml
with:
  go-version-file: go.mod
```

- Upgrade old action versions in `ko.yml` where practical (`actions/checkout@v4`, `actions/setup-go@v5`).

## Acceptance criteria

- All workflows use Go from `go.mod`.
- README build instructions match `go.mod`.
- Tag-triggered container publish works with Go 1.25.

### LINKED ISSUE #62: Regenerate README CLI help for Kong and document jq streaming behavior
## Problem

README usage text appears to be from the previous go-flags CLI output (`Application Options`, `Help Options`, `--sql=` formatting). The program now uses Kong struct tags for argument/flag/env/xor/enum handling.

The pasted help block also omits `--jq-input-mode`, even though the flag is present in the `opts` struct and documented elsewhere in README.

Default `--jq-input-mode=eager` materializes a full ResultSet via `resultset.Materialize`, which can consume large amounts of memory. The README mentions large result sets but should more prominently guide users toward lazy streaming.

## Verified locally

- README still shows go-flags-style help block (lines 24–56).
- `--jq-input-mode` exists in `main.go` but is missing from the pasted usage block.

## Suggested fix

Regenerate the README help block from current `execspansql --help`.

Document the memory behavior:

- default `--jq-input-mode=eager` materializes a full ResultSet
- `--jq-input-mode=lazy --filter='.rows[]'` streams rows
- filters such as `.rows | length`, indexing, slicing, nested `.rows`, or accessing `.stats` can materialize rows

Consider a golden test or `make update-readme-help` script so help drift is caught during PR review.

## Acceptance criteria

- README help matches current Kong output.
- README includes a streaming example for large result sets.
- README limitations section reflects current output formats (including `experimental_csv` if still supported).

### LINKED ISSUE #63: Add CLI-routing tests for timestamp bound, tracing, DML, and help output
## Problem

Existing unit/integration tests cover params, param files, CSV contracts, lazy jq semantics, row iterator behavior, and emulator-backed Spanner behavior well. CLI-routing edge cases are less directly tested.

## Suggested coverage

Add small seams and tests for:

- DML classification:
  - plain UPDATE/INSERT/DELETE
  - leading comments
  - SELECT remains read-only
- `--try-partition-query` uses the selected timestamp bound
- `--experimental-trace-project` preserves the configured timeout
- Kong help output includes all important flags (`--jq-input-mode`, etc.)
- `--log-grpc` safe logging behavior once redaction/gating is implemented

## Context

Related bugs tracked separately: #57 (tracing timeout), #58 (DML classification), #59 (log-grpc), and the partition-query timestamp bound issue.

## Acceptance criteria

- Tests do not require live Spanner except where emulator coverage is intentional.
- Help output drift is caught by a golden or focused assertion.

### PR DESCRIPTION
**Pull request by apstndb**
## Summary

- Change `--log-grpc` to an explicit `off|metadata|payload` mode so payload logging requires an intentional opt-in.
- Reject jq/raw/compact options on output paths that do not use them, and avoid compiling jq filters for CSV and partition-query checks.
- Add focused routing and validation tests for comment-prefixed DML, timestamp bounds, gRPC logging modes, and unsupported flag combinations.
- Refresh README help text, jq lazy-mode memory wording, partition-query timestamp guidance, gRPC payload caveats, Makefile targets, and Go workflow setup.

Closes #58.
Closes #59.
Closes #60.
Closes #61.
Closes #62.
Closes #63.

## Validation

- `git diff --check`
- `go test ./params/... ./jqresult/... ./resultset/...`
- `go test .`
- `make build`
- `make lint`
- `make test`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -298,26 +397,7 @@ func _main() error {
...
-	var m queryMode
-	switch {
-	case o.EnablePartitionedDML:
-		m = partitionedDML{}
-	case stmtkind.IsDMLLexical(query):
-		m = readWrite{}
-	default:
-		m = single{tb}
-	}
+	m := queryModeForQuery(query, o.EnablePartitionedDML, tb)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Addressed in 9b7689a: --try-partition-query and --read-timestamp now reject DML and partitioned-DML modes before client setup, with coverage for the DML and --enable-partitioned-dml cases.

File: main.go [hunk 2]
```diff
@@ -152,24 +149,115 @@ func readFileOrDefault(filename, s string) (string, error) {
...
+			return fmt.Errorf("--try-partition-query does not support jq filtering options")
+		}
+	}
+	return nil
+}
+
+func logGrpcClientOptions(logGrpcMode string) []option.ClientOption {
 	zapDevelopmentConfig := zap.NewDevelopmentConfig()
 	zapDevelopmentConfig.DisableCaller = true
 	zapLogger, _ := zapDevelopmentConfig.Build(zap.Fields())
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Addressed in 9b7689a: gRPC logging now falls back to zap.NewNop() if zap logger construction fails, with coverage for the fallback path.

File: flags_validation_test.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+			err:  "--read-timestamp cannot be used with DML statements",
+		},
+		{
+			name: "read_timestamp_rejects_partitioned_dml",
+			o: opts{
+				EnablePartitionedDML: true,
+				TimestampBound: struct {
+					Strong        bool   `name:"strong" xor:"timestamp" help:"Perform a strong query."`
+					ReadTimestamp string `name:"read-timestamp" xor:"timestamp" help:"Perform a query at the given timestamp. (micro-seconds precision)"`
+				}{ReadTimestamp: "2025-01-01T00:00:00Z"},
```
[the remark this answers is not in the document]
Comment (apstndb) [reply flags_validation_test.go]: Addressed in a4ac063: the duplicated tagged TimestampBound literals are gone. Directly omitting the tags in the opts composite literal does not compile, so the tests now use an optsWithReadTimestamp helper instead.

File: flags_validation_test.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+				ReadTimestamp string `name:"read-timestamp" xor:"timestamp" help:"Perform a query at the given timestamp. (micro-seconds precision)"`
+			}{ReadTimestamp: "2025-01-01T00:00:00Z"}},
+			mode: single{queryTimestamp},
+		},
+		{
+			name: "read_timestamp_rejects_dml",
+			o: opts{TimestampBound: struct {
+				Strong        bool   `name:"strong" xor:"timestamp" help:"Perform a strong query."`
+				ReadTimestamp string `name:"read-timestamp" xor:"timestamp" help:"Perform a query at the given timestamp. (micro-seconds precision)"`
+			}{ReadTimestamp: "2025-01-01T00:00:00Z"}},
```
[the remark this answers is not in the document]
Comment (apstndb) [reply flags_validation_test.go]: Addressed in a4ac063: the duplicated tagged TimestampBound literals are gone. Directly omitting the tags in the opts composite literal does not compile, so the tests now use an optsWithReadTimestamp helper instead.

File: main.go [hunk 5]
```diff
@@ -152,24 +149,143 @@ func readFileOrDefault(filename, s string) (string, error) {
...
+	if o.TryPartitionQuery {
+		if o.JqFilter != "" || o.JqFromFile != "" || o.JqRawOutput || o.CompactOutput || mode == jqresult.InputLazy {
+			return fmt.Errorf("--try-partition-query does not support jq filtering options")
+		}
+	}
+	return nil
+}
+
+func buildGrpcZapLogger(config zap.Config) *zap.Logger {
+	zapLogger, err := config.Build(zap.Fields())
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Addressed in a4ac063: buildGrpcZapLogger now calls config.Build() directly without the redundant empty zap.Fields() option.

File: flags_validation_test.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+			o:    opts{TryPartitionQuery: true, EnablePartitionedDML: true},
+			mode: partitionedDML{},
+			err:  "--try-partition-query cannot be combined with --enable-partitioned-dml",
+		},
+		{
+			name: "read_timestamp_allows_single_query",
+			o: opts{TimestampBound: struct {
+				Strong        bool   `name:"strong" xor:"timestamp" help:"Perform a strong query."`
+				ReadTimestamp string `name:"read-timestamp" xor:"timestamp" help:"Perform a query at the given timestamp. (micro-seconds precision)"`
+			}{ReadTimestamp: "2025-01-01T00:00:00Z"}},
```
[the remark this answers is not in the document]
Comment (apstndb) [reply flags_validation_test.go]: Addressed in a4ac063: the duplicated tagged TimestampBound literals are gone. Directly omitting the tags in the opts composite literal does not compile, so the tests now use an optsWithReadTimestamp helper instead.

File: main.go [hunk 7]
```diff
@@ -152,24 +149,143 @@ func readFileOrDefault(filename, s string) (string, error) {
...
+		}
+	}
+	if o.TimestampBound.ReadTimestamp != "" {
+		if _, ok := mode.(single); !ok {
+			if o.EnablePartitionedDML {
+				return fmt.Errorf("--read-timestamp cannot be combined with --enable-partitioned-dml")
+			}
+			return fmt.Errorf("--read-timestamp cannot be used with DML statements")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Addressed in 369e9ef: explicit --strong is now rejected for DML and partitioned-DML modes, with tests for both error paths.
